### PR TITLE
122 Fix Event Previews

### LINKED ISSUE #122: 500 error on preview event pages
Trying to preview a paged based on CoderedEventPage in the wagtail admin creates a 500 error. This is a persistent error we have seen in CodeRed CMS for some time, but it only affects certain page types. I believe it might have something to do with our `get_index_children()` method, or it could be entirely a problem in wagtail itself.

Recent stack trace:
```
File "/usr/local/lib/python3.7/site-packages/django/template/base.py" in _resolve_lookup
  829.                     current = current[bit]

During handling of the above exception ('EventPage' object is not subscriptable), another exception occurred:

File "/usr/local/lib/python3.7/site-packages/django/core/handlers/exception.py" in inner
  34.             response = get_response(request)

File "/usr/local/lib/python3.7/site-packages/django/core/handlers/base.py" in _get_response
  156.                 response = self.process_exception_by_middleware(e, request)

File "/usr/local/lib/python3.7/site-packages/django/core/handlers/base.py" in _get_response
  154.                 response = response.render()

File "/usr/local/lib/python3.7/site-packages/django/template/response.py" in render
  106.             self.content = self.rendered_content

File "/usr/local/lib/python3.7/site-packages/django/template/response.py" in rendered_content
  83.         content = template.render(context, self._request)

File "/usr/local/lib/python3.7/site-packages/django/template/backends/django.py" in render
  61.             return self.template.render(context)

File "/usr/local/lib/python3.7/site-packages/django/template/base.py" in render
  171.                     return self._render(context)

File "/usr/local/lib/python3.7/site-packages/django/template/base.py" in _render
  163.         return self.nodelist.render(context)

File "/usr/local/lib/python3.7/site-packages/django/template/base.py" in render
  937.                 bit = node.render_annotated(context)

File "/usr/local/lib/python3.7/site-packages/django/template/base.py" in render_annotated
  904.             return self.render(context)

File "/usr/local/lib/python3.7/site-packages/django/template/loader_tags.py" in render
  150.             return compiled_parent._render(context)

File "/usr/local/lib/python3.7/site-packages/django/template/base.py" in _render
  163.         return self.nodelist.render(context)

File "/usr/local/lib/python3.7/site-packages/django/template/base.py" in render
  937.                 bit = node.render_annotated(context)

File "/usr/local/lib/python3.7/site-packages/django/template/base.py" in render_annotated
  904.             return self.render(context)

File "/usr/local/lib/python3.7/site-packages/django/template/loader_tags.py" in render
  150.             return compiled_parent._render(context)

File "/usr/local/lib/python3.7/site-packages/django/template/base.py" in _render
  163.         return self.nodelist.render(context)

File "/usr/local/lib/python3.7/site-packages/django/template/base.py" in render
  937.                 bit = node.render_annotated(context)

File "/usr/local/lib/python3.7/site-packages/django/template/base.py" in render_annotated
  904.             return self.render(context)

File "/usr/local/lib/python3.7/site-packages/django/template/loader_tags.py" in render
  150.             return compiled_parent._render(context)

File "/usr/local/lib/python3.7/site-packages/django/template/base.py" in _render
  163.         return self.nodelist.render(context)

File "/usr/local/lib/python3.7/site-packages/django/template/base.py" in render
  937.                 bit = node.render_annotated(context)

File "/usr/local/lib/python3.7/site-packages/django/template/base.py" in render_annotated
  904.             return self.render(context)

File "/usr/local/lib/python3.7/site-packages/django/template/loader_tags.py" in render
  150.             return compiled_parent._render(context)

File "/usr/local/lib/python3.7/site-packages/django/template/base.py" in _render
  163.         return self.nodelist.render(context)

File "/usr/local/lib/python3.7/site-packages/django/template/base.py" in render
  937.                 bit = node.render_annotated(context)

File "/usr/local/lib/python3.7/site-packages/django/template/base.py" in render_annotated
  904.             return self.render(context)

File "/usr/local/lib/python3.7/site-packages/django/template/loader_tags.py" in render
  150.             return compiled_parent._render(context)

File "/usr/local/lib/python3.7/site-packages/django/template/base.py" in _render
  163.         return self.nodelist.render(context)

File "/usr/local/lib/python3.7/site-packages/django/template/base.py" in render
  937.                 bit = node.render_annotated(context)

File "/usr/local/lib/python3.7/site-packages/django/template/base.py" in render_annotated
  904.             return self.render(context)

File "/usr/local/lib/python3.7/site-packages/django/template/loader_tags.py" in render
  150.             return compiled_parent._render(context)

File "/usr/local/lib/python3.7/site-packages/django/template/base.py" in _render
  163.         return self.nodelist.render(context)

File "/usr/local/lib/python3.7/site-packages/django/template/base.py" in render
  937.                 bit = node.render_annotated(context)

File "/usr/local/lib/python3.7/site-packages/django/template/base.py" in render_annotated
  904.             return self.render(context)

File "/usr/local/lib/python3.7/site-packages/django/template/loader_tags.py" in render
  62.                 result = block.nodelist.render(context)

File "/usr/local/lib/python3.7/site-packages/django/template/base.py" in render
  937.                 bit = node.render_annotated(context)

File "/usr/local/lib/python3.7/site-packages/django/template/base.py" in render_annotated
  904.             return self.render(context)

File "/usr/local/lib/python3.7/site-packages/django/template/loader_tags.py" in render
  62.                 result = block.nodelist.render(context)

File "/usr/local/lib/python3.7/site-packages/django/template/base.py" in render
  937.                 bit = node.render_annotated(context)

File "/usr/local/lib/python3.7/site-packages/django/template/base.py" in render_annotated
  904.             return self.render(context)

File "/usr/local/lib/python3.7/site-packages/django/template/defaulttags.py" in render
  510.         values = {key: val.resolve(context) for key, val in self.extra_context.items()}

File "/usr/local/lib/python3.7/site-packages/django/template/defaulttags.py" in <dictcomp>
  510.         values = {key: val.resolve(context) for key, val in self.extra_context.items()}

File "/usr/local/lib/python3.7/site-packages/django/template/base.py" in resolve
  671.                 obj = self.var.resolve(context)

File "/usr/local/lib/python3.7/site-packages/django/template/base.py" in resolve
  796.             value = self._resolve_lookup(context)

File "/usr/local/lib/python3.7/site-packages/django/template/base.py" in _resolve_lookup
  837.                         current = getattr(current, bit)

File "/usr/local/lib/python3.7/site-packages/coderedcms/models/page_models.py" in most_recent_occurrence
  774.         noc = self.next_occurrence()

File "/usr/local/lib/python3.7/site-packages/eventtools/models.py" in next_occurrence
  152.             self.all_occurrences(from_date=from_date, to_date=to_date))

File "/usr/local/lib/python3.7/site-packages/eventtools/models.py" in all_occurrences
  271.         return self.get_related_occurrences().all_occurrences(

Exception Type: AttributeError at /admin/pages/1391/edit/preview/ Exception Value: 'FakeQuerySet' object has no attribute 'all_occurrences'
```

### PR DESCRIPTION
**Pull request by corysutyak**
Fixes #122 

I'm not entirely sure of the nature behind this issue.  Wagtail uses a `FakeQuerySet` in previews for some reason.  This `FakeQuerySet` appears to break with some eventtools apis as the `FakeQuerySet` isn't an exact copy of the model's `QuerySet`.  This is only an issue when we call `next_or_most_recent_occurrence` from template tags on previewed pages.  Without going down a miles deep rabbit hole, the solution is to manually calculate the next occurrence when the normal functionality breaks.  That solution is implemented in this PR.

There is another issue with the Event Landing page breaking on preview.  However, that is even harder to debug and unrelated to this specific issue.

### CODE REVIEW

File: coderedcms/blocks/content_blocks.py [hunk 1]
```diff
@@ -4,6 +4,7 @@
 """
 
 from django.utils.translation import ugettext_lazy as _
+from taggit.forms import TagField, TagWidget
```
Comment (vsalvino) [inline coderedcms/blocks/content_blocks.py]: Tag changes seem unrelated to this PR

File: coderedcms/models/page_models.py [hunk 2]
```diff
@@ -766,21 +766,28 @@ def upcoming_occurrences(self):
         """
         return self.query_occurrences(num_of_instances_to_return=10)
 
+
+    def _most_recent_occurrence(self):
```
Comment (vsalvino) [inline coderedcms/models/page_models.py]: This code could probably stay where it was in the `most_recent_occurrence()` function, I don't really see the need to split it out since it should never be called on its own.

File: coderedcms/models/page_models.py [hunk 3]
```diff
@@ -766,21 +766,28 @@ def upcoming_occurrences(self):
...
-            if len(aoc) > 0:
-                return aoc[-1] # last one in the list
-            return None
+        try:
+            noc = self.next_occurrence()
+            return noc if noc else self._most_recent_occurrence()
+        except AttributeError:
+            # Triggers when a preview is initiated on an EventPage because it uses a FakeQuerySet object.
+            # Here we manually compute the next_occurrence
+            occurrences = [e.next_occurrence() for e in self.occurrences.all()]
```
Comment (vsalvino) [inline coderedcms/models/page_models.py]: So the issue was that `self.next_occurrence()` breaks on preview, but `self.occurrences.all()[0].next_occurrence()` DOES work? I guess that sorta makes sense if there is a related_name that does not exist with the FakeQuerySet. I would be fine just showing the first occurrence on a preview, which would simplify and reduce the resources needed to generate a preview of this page.

File: coderedcms/models/page_models.py [hunk 4]
```diff
@@ -766,21 +766,28 @@ def upcoming_occurrences(self):
...
+    def _most_recent_occurrence(self):
+        aoc = []
+        for occurrence in self.occurrences.all():
+            aoc += [instance for instance in occurrence.all_occurrences()]
+        if len(aoc) > 0:
+            return aoc[-1] # last one in the list
+
     @property
-    def most_recent_occurrence(self):
+    def next_or_most_recent_occurrence(self):
```
Comment (vsalvino) [inline coderedcms/models/page_models.py]: I don't really see the need to rename this, especially given comment above.

**Comment by vsalvino** [comment]
I would also like to see the event landing page fixed as well... I wonder if we could do some sort of workaround where certain data is omitted (next occurrences for example) are omitted on the event landing preview page (or just sent as None)? Or if we can't show the list of events for some reason, just insert a "[event list here]" placeholder on the preview page, just so it doesn't 500 at least and we can figure out a better solution in the future.

**Comment by corysutyak** [comment]
It appears that the event landing page only breaks on preview when the display order is set to next occurrence.

**Comment by corysutyak** [comment]
The omitting won't be necessary for the event pages with the fix.

In the landing pages case, it has nothing to do with the content of the page.  It has to do with an auto generated validation form for the preview view to check whether the page is valid so it can then render.  For some reason, the display order field doesn't see next occurrence as a valid choice.  It can be fixed, but it would uproot multiple layers of abstraction.

**Comment by corysutyak** [comment]
PR moved to https://github.com/coderedcorp/coderedcms/pull/127
